core/src/labels/curvedLabel.cpp

```diff
@@ -0,0 +1,264 @@
+                                          _viewState.viewportSize, clipped);
+
+        if (clipped) { return false; }
+
+        sampler.add(sp);
+
+        if (!inside) {
+            if ((sp.x > min.x && sp.x < max.x &&
+                 sp.y > min.y && sp.y < max.y)) {
+                inside = true;
+            }
+        }
+    }
+
+    if (!inside || sampler.sumLength() < m_dim.x) {
+        sampler.clearPoints();
+        return false;
+    }
+
+    auto center = sampler.point(m_anchorPoint);
+    m_screenCenter = glm::vec2(center);
+
+    // Chord length for minimal ~120 degree inner angles (squared)
+    // sin(60)*2
+    const float sqDirLimit = powf(1.7f, 2);
+    // Range to check for angle changes
+    const float sampleWindow = 20;
+
+    float width = m_dim.x;
+    float start = sampler.point(m_screenAnchorPoint).z - width * 0.5f;
+
+    if (start < 0 || start + width > sampler.sumLength()) {
+        sampler.clearPoints();
+        return false;
+    }
+
+    // Sets current segment to the segment on which the text starts.
+    glm::vec2 p, r; // unused
+    sampler.sample(start, p, r);
+    size_t startSegment = sampler.curSegment();
+
+    for (size_t i = startSegment + 1; i < _transform.size(); i++) {
+        float currLength = sampler.point(i).z;
+        glm::vec2 dir = sampler.segmentDirection(i);
+
+        // Go back within window to check for hard direction changes
+        for (int k = i - 1; k >= int(startSegment); k--) {
+            if (glm::length2(sampler.segmentDirection(k) + dir) < sqDirLimit) {
+                sampler.clearPoints();
+                return false;
+            }
+            if (sampler.point(k).z < sampler.point(i).z - sampleWindow) {
+                break;
+            }
+        }
+        if (currLength > start + width) {
+            break;
+        }
+    }
+
+    return true;
+}
+
+void CurvedLabel::obbs(ScreenTransform& _transform, OBBBuffer& _obbs) {
+
+    glm::vec2 dim = m_dim - m_options.buffer;
+
+    if (m_occludedLastFrame) { dim += Label::activation_distance_threshold; }
+
+    // TODO: Remove - Only for testing
+    if (state() == State::dead) { dim -= 4; }
+
+    float width = dim.x;
+    LineSampler<ScreenTransform> sampler { _transform };
+
+    auto center = sampler.point(m_screenAnchorPoint).z;
+    //auto center = sampler.sumLength() * 0.5;
+
+    auto start = center - width * 0.5f;
+
+    glm::vec2 p1, p2, rotation;
+    sampler.sample(start, p1, rotation);
+
+    float prevLength = start;
+
+    int count = 0;
+    for (size_t i = sampler.curSegment()+1; i < _transform.size(); i++) {
+
+        float currLength = sampler.point(i).z;
+        float segmentLength = currLength - prevLength;
+        count++;
+
+        if (start + width > currLength) {
+            p2 = glm::vec2(sampler.point(i));
+
+            rotation = sampler.segmentDirection(i-1);
+            _obbs.append({(p1 + p2) * 0.5f, rotation, segmentLength, dim.y});
```

Comment:
Label offset for curved labels needs the be applied to the polyline in updateScreenTransform. I'll do this in the next PR.